cpp/src/knn/knn_opg_common.cu

```diff
@@ -238,9 +315,15 @@ void reduce(raft::handle_t &handle, std::vector<Matrix::Data<T> *> *out,
                                        indices, cur_batch_size, idxRanks.size(),
                                        k, stream, trans.data());
 
-  perform_local_operation(output, indices, res.data(), cur_batch_size, k,
-                          n_outputs, handle, probas_only, &probas_with_offsets,
-                          uniq_labels, n_unique);
+  device_buffer<T> merged_outputs_b(alloc, stream,
+                                    n_outputs * cur_batch_size * k);
+  T *merged_outputs = merged_outputs_b.data();
+  merge_outputs(merged_outputs, indices, res.data(), res_I.data(),
```

Comment:
I notice we do a `knn_merge_parts` above but then we run this `merged_outputs`, which seems to be a (much less parallel) k-selection. The code gets a little hard to follow because the function signatures have grown so large. Can you explain the intention of the `merge_outputs` and the corresponding kernel?

Reply:
The original array of outputs/labels is of size `n_ranks` * `batch_size` * `n_outputs`. We would like to get the merged array of outputs/labels of size `batch_size` * `n_outputs`. This can be obtained by matching the merged indices that we got from `knn_merge_parts` with their original position in the unmerged indices array. With this positions we can get the associated output in the unmerged array of outputs and write it back to the merged array of outputs.

`outputs` : merged outputs
`knn_indices`: merged indices
`unmerged_outputs` : unmerged outputs
`unmerged_knn_indices`: unmerged indices
`offsets`: array splitting the partitions, making it possible to identify the origin partition of an nn index
`parts_to_ranks`: get rank index from index partition index, informative to find positions as the unmerged arrays are built so that ranks are in order (unlike partitions)
`nearest_neighbors`: number of nearest neighbors to look for in query
`n_outputs`: number of targets
`n_labels`: number of outputs/labels to write (`batch_size` * `n_outputs`)
`n_parts`: number of index partitions
`n_ranks`: number of index ranks

Then comes the question of parallelizing the two loops:
- The first one `for (int k = 0; k < nearest_neighbors; k++)` exponentially often stops on the few first iterations as the nn index that was selected in the merged array is probably the closest (appearing first) in the unmerged array. Parallelizing this loop would increase global memory accesses and would probably not be worth it.
- The second one `for (int o = 0; o < n_outputs; o++)` is just about writing back at the right location in the merged array of outputs/labels. Likewise running the kernel asynchronously with each run taking care of an output/target will result in much more global memory accesses to find the right positions.

To sum it up it's a tradeoff between the level of parallelization and memory accesses. What do you think is preferable?